File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/lists_setIndex.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception
import java.util.ArrayList

suspend fun Piper.executeListsSetIndex(statement: JSONObject, onError: (e: Exception) -> Unit) {
    try {
        Log.d("hello", "execute_lists_setIndex")
        val pointer = statement.getJSONObject("pointer")
        val list = executeStatement(pointer["list"].toString()) as ArrayList<Any?>
        val index = executeStatement(pointer["index"].toString()).toString().toInt()
        val value = executeStatement(pointer["value"].toString())
        val mode = pointer["mode"].toString()

        when (pointer["where"].toString()) {
            "FROM_START" -> when (mode) {
                "SET" -> list[index - 1] = value
                "INSERT" -> list.add(index - 1, value)
            }
            "FROM_END" -> when (mode) {
                "SET" -> list[list.size - index] = value
                "INSERT" -> list.add(list.size - index, value)
            }
            "FIRST" -> when (mode) {
                "SET" -> list[0] = value
                "INSERT" -> list.add(0, value)
            }
            "LAST" -> when (mode) {
                "SET" -> list[list.lastIndex] = value
                "INSERT" -> {
                    list.add(list.size, value)
                }
            }
            "RANDOM" -> when (mode) {
                "SET" -> list[(0 until list.size).random()] = value
                "INSERT" -> list.add((0 until list.size).random(), value)
            }
        }
    } catch (e: Exception) {
        onError(e)
    }
}

File: piper/extensions/chrome/src/engine/statements/lists_setIndex.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let list = await this.executeStatement(stringify(pointer.list))
        let index = await this.executeStatement(stringify(pointer.index))
        index = parseInt(index)
        let value = await this.executeStatement(stringify(pointer.value))
        let mode = stringify(pointer.mode)
        let where = stringify(pointer.where)

        if (where === "FROM_START") {
            if (mode === "SET") list[index - 1] = value
            else if (mode === "INSERT") list.splice(index - 1, 0, value)
        } else if (where === "FROM_END") {
            if (mode === "SET") list[list.length - index] = value
            else if (mode === "INSERT") list.splice(list.length - index, 0, value)
        } else if (where === "FIRST") {
            if (mode === "SET") list[0] = value
            else if (mode === "INSERT") list.splice(0, 0, value)
        } else if (where === "LAST") {
            if (mode === "SET") list[list.length - 1] = value
            else if (mode === "INSERT") list.splice(list.length - 1, 0, value)
        } else if (where === "RANDOM") {
            let rand = Math.floor(Math.random() * list.length)
            if (mode === "SET") list[rand] = value
            else if (mode === "INSERT") list.splice(rand, 0, value)
        }
    } catch (e) {
        onError(e)
    }
}
